Improvement/dropdown css only

AndyOGo opened the pull request:
Hey all

This aims to fix #466 

You can check it out in the dropdown section - CSS-only.

This solution is utilizing floats and negative margins to the rescue.

---

The current Dropdown is utilising JS to make it stylable.
We can improve this with CSS appearance rule on window-based elements like input, select, etc.

Originally I developed such a fancy but JS-less dropdown for [m.tutti.ch](http://m.tutti.ch) (*IMPORTANT*: visit it with a mobile User Agent)

![dropdown_m tutti ch mov](https://cloud.githubusercontent.com/assets/914443/15715242/00f991fa-281e-11e6-948f-e05ba56d050a.gif)

This concepts had to meet the following prerequisites:
 - Allow for icons before and after the dropdown`s label
 - Allow for fancy styling with CSS3
 - avoid `z-index` (cause of `z-index` issues with old iOS devices)
 - gracefull fallback to OS specific rendering
 - instant rendering (no Flash caused by JS)
 - adding a X-Icon to reset the dropdown (JS-enhanced, optional)
 - Expects to fit the containing block's size (fluid and 100% width)
 - User should use the native platform specific UI to select an option

*Issues*
 - older Firefox version didn't remove their native arrows in spite of -moz-appearance: none (can be fixed)
 - `<select>` elements are window-elements and hence not reliable to style with rules like `line-height` and `vertiacal-align: middle` (can be fixed with padding)

*Important Refs*:
 - [CanIUse](http://caniuse.com/#feat=css-appearance)
 - [CSS-Tricks](https://css-tricks.com/almanac/properties/a/appearance/)
 - [IE supports -webkit- oo](https://msdn.microsoft.com/en-us/library/dn793580(v=vs.85).aspx)
 - [fix IE Selectbox arrows](https://msdn.microsoft.com/en-us/library/hh771821(v=vs.85).aspx)
 - [fix IE active value blue bg](https://msdn.microsoft.com/en-us/library/windows/apps/hh465820.aspx)
 - [StackOverflow](http://stackoverflow.com/questions/1895476/how-to-style-a-select-dropdown-with-css-only-without-javascript)
 - [FF double `text-indent` issue](http://stackoverflow.com/questions/28108434/firefox-select-element-doubles-text-indent-value)

elessar-ch commented:
Looks great to me and should make it easier to implement this component in whatever frameworks/libraries our users use.

Also works in IE11 and apparently IE10:
![dropdown2_ie11](https://cloud.githubusercontent.com/assets/7779684/15717132/e1e948e8-2825-11e6-926a-8e7000db77bb.gif)

IE9 just has a little styling issue when you selected an item. When you tab or click away the state returns to normal though:
![dropdown2_ie9](https://cloud.githubusercontent.com/assets/7779684/15717242/6452239a-2826-11e6-8d24-6085818c3a0f.gif)

davidknezic commented:
AWESOME @AndyOGo!

davidknezic commented:
I could even imagine making this the default `dropdown` block and releasing it in a new breaking change version! What do you think @sventschui @elessar-ch @martinoss?

AndyOGo commented:
@davidknezic Thanks :)

Yeah that would be awesome, to make it final we should really test if `vertical-align: middle` in combination with `line-height` is reliable in the browsers we target. The days back at tutti, it wasn't.

And if I remember correctly firefox hat a [double `text-indent` issue](http://stackoverflow.com/questions/28108434/firefox-select-element-doubles-text-indent-value) on window-elements.

AndyOGo commented:
Oooh my hooooly madness, this double `text-indent` bug is still there. check this [fiddle](http://jsfiddle.net/k92dvxte/1/) in newest firefox.

sventschui commented:
Thanks a lot!

- I'd not update the dropdown block. Breaking changes will make it harder for people to upgrade
- Shall it be named `dropdown2` or `dropown-2`?

AndyOGo commented:
I just commited a few updates:

 - added a mixin to target firefox
 - added a mixin to target IE 10+
 - utilized `text-indent`
 - regular and large version
 - code cleanup

paddings still need to be reset properly

elessar-ch commented:
I agree with @davidknezic: we should make it our default. 
And I agree with @sventschui: we should not replace the dropdown block.
I would name it dropdown2 similar to the button2 we already have in the docs for the global page.

We should also add the large and condensed modifiers. Then put the description for the js dropdowns in a deprecated section or something on the docs page.

AndyOGo commented:
@elessar-ch just added --condensed modifer

@elessar-ch @davidknezic @sventschui 
I agree to avoid breaking existing stuff and name it dropdown-2.
I added a punch of fixes, to deal with [form styling nightmare](https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Forms/Advanced_styling_for_HTML_forms#Dealing_with_the_select_nightmare) 
Before this goes into production, we need to test it well. Primarily on desktop browsers, because it is successfully in production at m.tutti.ch since 2+ years (which is not meant for desktop).

AndyOGo commented:
Update, forms will change a lot. Don't know if this is still applicable.

sventschui commented:
You should definitely talk to the guys building the new form components. Should have a css only select/dropdown there

AndyOGo commented:
@sventschui 
I did, unfortunately they get pretty fancy, including animating different states.
Checkout the [material-inspired branch](/axa-ch/style-guide/tree/feature/material-inspired-forms-v2)

AndyOGo commented:
Update, since we got a new issue #495 regarding `dropdown` in IE 11.
This gets traction again.
